### PR TITLE
HLRC: Add Lifecycle delete to the HLRC

### PR DESCRIPTION
**Pull request by AthenaEryma**
Adds support for Lifecycle deletion to the Java High-Level Rest Client.
Also includes some stopgap measures to support using the new Client
package structure which should be replaced once the structure
refactoring is complete.

Relates to #33100

### CODE REVIEW

File: client/rest-high-level/src/test/java/org/elasticsearch/client/indexlifecycle/DeleteLifecycleRequestTests.java
Comment (dakrone) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/indexlifecycle/DeleteLifecycleRequestTests.java]: Can you split this into two different tests rather than using the randomness here? I don't think it buys us anything

**Review by dakrone — APPROVED** [review]
This LGTM assuming @hub-cap is good with the `MasterTimeoutRequest` change

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/DeleteLifecycleRequest.java
Comment (jakelandis) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/DeleteLifecycleRequest.java]: The transport based APIs extend `AcknowledgedRequest` which has a `timeout` and `ackTimeout` in addition to the `masterNodeTimeout`.

Curious if is this is intentional ?

The javadoc for `AcknowledgedRequest` states `Facilitates consistency across different api.` ... is that something we should carry forward while still divorcing ourselves from the transport request objects ?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/DeleteLifecycleRequest.java
Comment (jakelandis) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/DeleteLifecycleRequest.java]: Should this be Delete`Index`LifecycleRequest ?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/DeleteLifecycleRequest.java
Comment (gwbrown) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/DeleteLifecycleRequest.java]: I based `MasterTimeoutRequest` on `MasterNodeRequest`, rather than `AcknowledgedRequest`, which doesn't have an `ackTimeout`.  It's possible I should be using `AcknowledgedRequest` instead - the difference wasn't immediately clear.

So, to answer your question: It was intentional, but it's possible I intended the wrong thing.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/DeleteLifecycleRequest.java
Comment (gwbrown) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/DeleteLifecycleRequest.java]: From what I saw, the classes with names including `IndexLifecycle` seem to relate to index<->lifecycle policy relationships (i.e. lifecycle `A` applies to index `B`), but ones with just `Lifecycle` relate to the management of lifecycles.  If that's correct, this class is named correctly.

@dakrone, can you confirm?

Comment (dakrone) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/DeleteLifecycleRequest.java]: I think this should be `DeleteLifecyclePolicyRequest` to indicate that the policy is going to be removed, and then `DeleteIndexLifecycleRequest` is for when we are removing a lifecycle from an index

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/DeleteLifecycleRequest.java
Comment (gwbrown) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/DeleteLifecycleRequest.java]: I talked to @hub-cap on Slack and his answer was that these two timeouts should be based on the parameters that the requests can take.  To determine this, there should be a `Rest<whatever>Action` class which defines those parameters.  In this case, `RestDeleteLifecycleAction` (see [here](https://github.com/elastic/elasticsearch/blob/57835452228ba03238ef39e19829a67581689843/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/indexlifecycle/action/RestDeleteLifecycleAction.java#L35)) defines both `timeout` and `master_timeout`, so this class should include both.

(Just waiting on some more feedback to make this change)

Comment (jakelandis) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/DeleteLifecycleRequest.java]: @gwbrown Thanks for following up. I am working a similar issue and will follow your lead here.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/IndexLifecycleClient.java
Comment (colings86) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/IndexLifecycleClient.java]: In the same spirit as @dakrone's comment about renaming the request object to `DeleteLifecyclePolicyRequest` I think we should rename these methods to `deleteLifecyclePolicy` and `deleteLifecyclePolicyAsync`?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java
Comment (colings86) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: Rename to `deleteLifecyclePolicy`?

**Review by colings86 — COMMENTED** [review]
I left a couple of minor comments

**Comment by gwbrown** [comment]
Just had a brief meeting with @hub-cap - the result of that meeting is that there are some not-well-understood requirements around exactly which timeouts need to/can be shared between request classes.  Baz is going to figure those requirements out when he has bandwidth to do so.  In the meantime the abstract request class in this PR will change to include both `timeout` and `master_timeout` to allow moving forward with this and other ILM HLRC PR's and we'll consolidate to the final hierarchy later - it ought to be pretty straightforward.

**Comment by gwbrown** [comment]
I lied - @hub-cap got the base class into `master` very quickly (thanks!) so this PR has transitioned to using that base class.

I've made all the changes suggest so far, so any last comments on this? If not I'll merge about EOD today, assuming CI passed.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java
Comment (dakrone) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: Yep, +1 to rename this `deleteLifecyclePolicy`

File: client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/DeleteLifecyclePolicyRequest.java
Comment (dakrone) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/indexlifecycle/DeleteLifecyclePolicyRequest.java]: Can you also rename these `getLifecyclePolicy`?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/indexlifecycle/DeleteLifecyclePolicyRequestTests.java
Comment (dakrone) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/indexlifecycle/DeleteLifecyclePolicyRequestTests.java]: For the future, you can use

```java
expectThrows(IllegalArgumentException.class, () -> new DeleteLifecyclePolicyRequest(randomFrom("", null)));
```

Instead of having to do the try/fail/catch check

**Review by dakrone — CHANGES_REQUESTED** [review]
I left comments about some renames

**Comment by hub-cap** [comment]
@gwbrown its also been backported as of about an hour ago. So you should be good to go for any 6.5 merging as well.

**Review by dakrone — APPROVED** [review]
LGTM

**Pull request merged**
